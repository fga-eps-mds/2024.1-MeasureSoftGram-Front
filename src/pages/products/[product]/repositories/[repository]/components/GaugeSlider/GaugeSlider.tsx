import React, { useState } from "react";
import { Box, Slider } from "@mui/material";

interface GaugeSliderProps {
  initialValues: number[];
  min: number;
  max: number;
  values: number[];
  setValues: (newValues: number[]) => void;
  step: number;
}

function GaugeSlider(props: GaugeSliderProps) {
  const {
    initialValues,
    min,
    max,
    values,
    setValues,
    step
  } = props;

  const [perc, setPerc] = useState(
    initialValues.map((val: number) => (val / max) * 100)
  );

  const onChange = (e: Event, tValues: number | number[]) => {
    if (Array.isArray(tValues)) {
      const [minVal, maxVal] = tValues;
      if (maxVal > minVal && maxVal !== minVal && minVal > 0 && maxVal < 1) {
        setValues(tValues);
        setPerc(tValues.map((val: number) => (val / max) * 100));
      }
    }
  };

  return (
    <Box
      sx={{
        width: "100%",
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center'
      }}
    >
      <Slider
        data-testid="gauge-slider-id"
        disableSwap
        sx={{
          width: '90%',
          "& .MuiSlider-track": {
            background: "#f1c40f",
            border: 0
          },
          "& .MuiSlider-thumb": {
            [`&:nth-of-type(${1}n)`]: {
              background: "#313131",
              "& span": {
                background: "#313131",
                borderRadius: '4px',
              }
            },
            [`&:nth-of-type(${2}n)`]: {
              background: "#313131",
              "& span": {
                background: "#313131",
                borderRadius: '4px',
              }
            }
          },
          "& .MuiSlider-mark": {
            background: "none"
          },
          "& .MuiSlider-rail": {
            opacity: 100,
            border: 0,
            background: `linear-gradient(to right, #e74c3c 0% ${perc[0]}%, #f1c40f ${perc[0]}% ${perc[1]}%, #07bc0c ${perc[1]}% 100%)`
          }
        }}
        valueLabelDisplay="auto"
        value={values}
        min={min}
        max={max}
        scale={(x) => (x)}
        marks={[
          { value: min, label: min },
          ...perc.map((val: any) => {
            let label;
            if (val === 0) {
              label = min;
            } else if (val === 1) {
              label = max;
            } else {
              label = undefined;
            }
            return { value: val, label };
          }),
          { value: max, label: max }
        ]}
        onChange={onChange}
        step={step}
      />
    </Box>
  );
}

export default GaugeSlider;
