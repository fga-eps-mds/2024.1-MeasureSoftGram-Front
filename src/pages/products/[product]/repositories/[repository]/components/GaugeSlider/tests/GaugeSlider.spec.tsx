import React, { useState } from 'react';
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom';
import GaugeSlider from '../GaugeSlider';


describe('GaugeSlider', () => {
  it('should update values and percentage on change', async () => {
    const setValuesMock = jest.fn();

    const eventMock = {} as React.ChangeEvent<{}>;

    const { getByTestId } = render(
      <GaugeSlider
        initialValues={[0.33, 0.66]}
        min={0}
        max={1}
        values={[0.33, 0.66]}
        setValues={setValuesMock}
        step={0.01}
      />
    );

    const sliderContainer = getByTestId('gauge-slider-id');
    expect(sliderContainer).toBeTruthy()

    const sliderInput = await sliderContainer.querySelector('input[type="range"]');
    fireEvent.change(sliderInput!, { target: { value: [0.5] } });

    expect(setValuesMock).toHaveBeenCalledWith([0.5, 0.66]);

  });

  it('should match snapshot', () => {

    const setValuesMock = jest.fn();

    const eventMock = {} as React.ChangeEvent<{}>;

    const tree = render(
      <GaugeSlider
        initialValues={[0.33, 0.66]}
        min={0}
        max={1}
        values={[0.33, 0.66]}
        setValues={setValuesMock}
        step={0.01}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
