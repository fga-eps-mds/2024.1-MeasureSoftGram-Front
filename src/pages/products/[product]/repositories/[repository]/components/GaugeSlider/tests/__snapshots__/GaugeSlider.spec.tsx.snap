// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GaugeSlider should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-12auco2"
      >
        <span
          class="MuiSlider-root MuiSlider-marked MuiSlider-colorPrimary MuiSlider-sizeMedium css-dbrczr-MuiSlider-root"
          data-testid="gauge-slider-id"
        >
          <span
            class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
          />
          <span
            class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
            style="left: 33%; width: 33%;"
          />
          <span
            class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
            data-index="0"
            style="left: 0%;"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-markLabel MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
            data-index="0"
            style="left: 0%;"
          >
            0
          </span>
          <span
            class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
            data-index="1"
            style="left: 100%;"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-markLabel MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
            data-index="1"
            style="left: 100%;"
          >
            1
          </span>
          <span
            class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-eg0mwd-MuiSlider-thumb"
            data-index="0"
            style="left: 33%;"
          >
            <input
              aria-orientation="horizontal"
              aria-valuemax="1"
              aria-valuemin="0"
              aria-valuenow="0.33"
              data-index="0"
              max="1"
              min="0"
              step="0.01"
              style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
              type="range"
              value="0.33"
            />
            <span
              aria-hidden="true"
              class="MuiSlider-valueLabel css-nnid7-MuiSlider-valueLabel"
            >
              <span
                class="MuiSlider-valueLabelCircle"
              >
                <span
                  class="MuiSlider-valueLabelLabel"
                >
                  0.33
                </span>
              </span>
            </span>
          </span>
          <span
            class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-eg0mwd-MuiSlider-thumb"
            data-index="1"
            style="left: 66%;"
          >
            <input
              aria-orientation="horizontal"
              aria-valuemax="1"
              aria-valuemin="0"
              aria-valuenow="0.66"
              data-index="1"
              max="1"
              min="0"
              step="0.01"
              style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
              type="range"
              value="0.66"
            />
            <span
              aria-hidden="true"
              class="MuiSlider-valueLabel css-nnid7-MuiSlider-valueLabel"
            >
              <span
                class="MuiSlider-valueLabelCircle"
              >
                <span
                  class="MuiSlider-valueLabelLabel"
                >
                  0.66
                </span>
              </span>
            </span>
          </span>
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-12auco2"
    >
      <span
        class="MuiSlider-root MuiSlider-marked MuiSlider-colorPrimary MuiSlider-sizeMedium css-dbrczr-MuiSlider-root"
        data-testid="gauge-slider-id"
      >
        <span
          class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
        />
        <span
          class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
          style="left: 33%; width: 33%;"
        />
        <span
          class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index="0"
          style="left: 0%;"
        />
        <span
          aria-hidden="true"
          class="MuiSlider-markLabel MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index="0"
          style="left: 0%;"
        >
          0
        </span>
        <span
          class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index="1"
          style="left: 100%;"
        />
        <span
          aria-hidden="true"
          class="MuiSlider-markLabel MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
          data-index="1"
          style="left: 100%;"
        >
          1
        </span>
        <span
          class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-eg0mwd-MuiSlider-thumb"
          data-index="0"
          style="left: 33%;"
        >
          <input
            aria-orientation="horizontal"
            aria-valuemax="1"
            aria-valuemin="0"
            aria-valuenow="0.33"
            data-index="0"
            max="1"
            min="0"
            step="0.01"
            style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
            type="range"
            value="0.33"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-valueLabel css-nnid7-MuiSlider-valueLabel"
          >
            <span
              class="MuiSlider-valueLabelCircle"
            >
              <span
                class="MuiSlider-valueLabelLabel"
              >
                0.33
              </span>
            </span>
          </span>
        </span>
        <span
          class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-eg0mwd-MuiSlider-thumb"
          data-index="1"
          style="left: 66%;"
        >
          <input
            aria-orientation="horizontal"
            aria-valuemax="1"
            aria-valuemin="0"
            aria-valuenow="0.66"
            data-index="1"
            max="1"
            min="0"
            step="0.01"
            style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
            type="range"
            value="0.66"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-valueLabel css-nnid7-MuiSlider-valueLabel"
          >
            <span
              class="MuiSlider-valueLabelCircle"
            >
              <span
                class="MuiSlider-valueLabelLabel"
              >
                0.66
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
